### ISSUE TITLE
Clipboard copying quirks

### ISSUE THREAD

**Issue by pane** [zathras/jrpn#83]
**JRPN 16C or 15C?**
15C

**Describe the bug**
While using Clipboard for copy and paste some errors happen

**Debug Log**
{"version":1,"modelName":"15C","settings":{"menuEnabled":true,"windowEnabled":true,"longNumbers":0,"euroComma":false,"showAccelerators":false,"systemOverlaysDisabled":false,"orientation":0,"keyFeedback":0},"displayMode":"x4","integerSignMode":"2","wordSize":56,"stack":["0","0","0","0"],"lastX":"0","flags":[false,false,false,false,false,false,false,false,false,false],"memory":{"storage":"000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000","program":{"lines":0,"currentLine":0,"returnStack":[0,0,0,0,0,0,0],"returnStackPos":-1},"I":"0"},"debugLog":{"initialState":{"version":1,"modelName":"15C","settings":{"menuEnabled":true,"windowEnabled":true,"longNumbers":0,"euroComma":false,"showAccelerators":false,"systemOverlaysDisabled":false,"orientation":0,"keyFeedback":0},"displayMode":"x4","integerSignMode":"2","wordSize":56,"stack":["0","0","0","0"],"lastX":"0","flags":[false,false,false,false,false,false,false,false,false,false],"memory":{"storage":"000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000","program":{"lines":0,"currentLine":0,"returnStack":[0,0,0,0,0,0,0],"returnStackPos":-1},"I":"0"},"numRegisters":20,"resultMatrix":0,"matrices":[{"columns":0,"values":[],"rowSwaps":null},{"columns":0,"values":[],"rowSwaps":null},{"columns":0,"values":[],"rowSwaps":null},{"columns":0,"values":[],"rowSwaps":null},{"columns":0,"values":[],"rowSwaps":null}],"lastRandom":0.010973265189577974},"keys":[1,2,3,4,5,6,7,8,9,-2,124]},"endDisplay":" 0.0000     ","numRegisters":20,"resultMatrix":0,"matrices":[{"columns":0,"values":[],"rowSwaps":null},{"columns":0,"values":[],"rowSwaps":null},{"columns":0,"values":[],"rowSwaps":null},{"columns":0,"values":[],"rowSwaps":null},{"columns":0,"values":[],"rowSwaps":null}],"lastRandom":0.010973265189577974}

**To Reproduce**
Firefox version opened via https://jrpn.jovial.com/run15/index.html
Reset all and enter with keyboard
123456789
Copy to Clipboard
CLx
Paste from clipboard
Error message: Empty clipboard 

**Expected behavior**
Clipboard copy paste works

**Platform:**
Win system is set to enter decimal point by entering comma key "," (European standard, specifically Czech keyboard and environment although system is in English)

**Additional context**
Notes:
Ctrl V done on the line below inside the quotes shows the content of the clipboard
" 123,456,789"
My Win system is set to enter decimal point by pressing comma "," (European standard, specifically Czech keyboard and environment although system is in English)
Microsoft Edge returns error message: Number format error
Edge accepts Paste from clipboard if the content is:
" 123456789"
Firefox still returns error: Empty clipboard if the content is even without trailing space:
"123456789"
Problems with entering the program via Clipboard as well
Copying also from Win app not working

Please let me know if I should try some more action/setting.

KR
P

**Comment by zathras** [comment]
For the first bug, it was just a matter of stripping out unwanted characters.  
```
   Value? tryParse(String s, NumStatus m) {
+    s = s.trim().replaceAll(',', '');
     double? d = double.tryParse(s);
```
It seems Dart's double.tryParse() isn't so forgiving :-)

About Firefox, I was able to reproduce the behavior.  On Firefox, I get "empty clipboard."  On Chrome, it asks me if I want to grant the application permission to read the clipboard, and if I say "yes," it works.  I think this is just a problem with Flutter on Firefox -- it could be a Flutter bug, a Firefox bug, or a feature that Firefox hasn't implemented.

**Comment by pane** [comment]
Thank you for your constant improvements. 
I was going to try this fix in Chrome since I thought deployment #320 pushed latest changes to version accessible via https://jrpn.jovial.com/run15/index.html (or am I still running previous version - showing 2.1.12.beta2) but still getting Number format error
Would that fix of removal "," work if the radix is set to coma?
Thank you again. KR P

**Comment by zathras** [comment]
About the web version:  That only changes if I do a rebuild, which I usually only do when releasing.  I'm in the process of doing a release now, so I'll go ahead and build the web version.

Copy/paste wasn't doing anything with the comma setting -- it was only changing how the LCD displayed things.  But I probably should swap "," and "." for copy/paste in Euro mode...  See next push.

**Comment by zathras** [comment]
Published in 2.1.12

**Comment by pane** [comment]
> But I probably should swap "," and "." for copy/paste in Euro mode... See next push.

Would it be also possible to implement to accept the coma "," as a decimal point entry alongside with the dot ".", please? My "decimal point" key on the PC NumKeypad part is mapped to "," and does not register as the decimal point "." (found out when doing my Taxes and was not getting right figures ). I hope this will not make mess with other languages implementations. I am slowly looking to the source but since I am not a programmer I am not able to fork and build it yet.

Thank you,
KR
P
